### PR TITLE
Jh/full scanner

### PR DESCRIPTION
**Pull request by jibben**
I think this is a fully formed scanner.mll file

### CODE REVIEW

File: src/scanner.mll [hunk 1]
```diff
@@ -3,18 +3,29 @@
...
+
+    (* comma, dot, semi *)
+    | '.'                   { DOT }
+    | ','                   { COMMA }
+    | ';'                   { SEMI }
+
+    (* logic *)
     | "=="                  { EQ }
     | "!="                  { NEQ }
     | '<'                   { LT }
```
Comment (vlad-scherbich) [inline src/scanner.mll:31]: What about: 
GT, LEQ

File: src/scanner.mll [hunk 2]
```diff
@@ -23,30 +34,68 @@ rule token = parse
...
+    | "bool"                { BOOL }
+    | "unit"                { UNIT }
+
+    (* non-primitive types *)
+    | "string"              { STR }
+    | "maybe"               { OPTION_MAYBE }
+    | "none"                { OPTION_NONE }
+    | "some"                { OPTION_SOME }
     | "list"                { LIST }
-    | "MAYBE"               { OPTION_MAYBE }
```
Comment (vlad-scherbich) [inline src/scanner.mll:46]: I wanted to ask before: why are MAYBE, NONE, SOME capitalized?

File: src/scanner.mll [hunk 3]
```diff
@@ -23,30 +34,68 @@ rule token = parse
...
+    | '|'                   { BITWISE_OR }
+    | '^'                   { BITWISE_XOR }
+    | '~'                   { BITWISE_NOT }
+    | ">>"                  { BITSHIFT_RIGHT }
+    | "<<"                  { BITSHIFT_LEFT }
+
+    (* functions *)
+    | ':'                   { FUNCTION_TYPE }
+    | "->"                  { ARROW }
+    | "=>"                  { FUNCTION }
```
Comment (vlad-scherbich) [inline src/scanner.mll:51]: `func` missing

File: src/scanner.mll [hunk 4]
```diff
@@ -3,18 +3,29 @@
...
     | '['                   { LBRACKET }
     | ']'                   { RBRACKET }
+
+    (* math *)
     | '+'                   { PLUS }
     | '-'                   { MINUS }
     | '*'                   { TIMES }
     | '/'                   { DIVIDE }
-    | ','                   { COMMA }
     | '='                   { ASIGN }
```
Comment (vlad-scherbich) [inline src/scanner.mll]: nit: `ASSIGN`

File: src/scanner.mll [hunk 5]
```diff
@@ -3,18 +3,29 @@
 rule token = parse
     | [' ' '\t' '\r' '\n']  { token lexbuf }
     | "/*"                  { comment lexbuf }
```
Comment (vlad-scherbich) [inline src/scanner.mll:5]: `/*` missing?

File: src/scanner.mll [hunk 6]
```diff
@@ -3,18 +3,29 @@
...
     | '{'                   { LBRACE }
     | '}'                   { RBRACE }
     | '['                   { LBRACKET }
     | ']'                   { RBRACKET }
+
+    (* math *)
     | '+'                   { PLUS }
     | '-'                   { MINUS }
     | '*'                   { TIMES }
     | '/'                   { DIVIDE }
```
Comment (vlad-scherbich) [inline src/scanner.mll:19]: Do we need a separate rule for `//` (line comment)?

File: src/scanner.mll (hunk 1 above)
Comment (jibben) [reply src/scanner.mll:31]: They are there?

File: src/scanner.mll (hunk 2 above)
Comment (jibben) [reply src/scanner.mll:46]: fixed

File: src/scanner.mll (hunk 5 above)
Comment (jibben) [reply src/scanner.mll:5]: It's right there on line 5..?

File: src/scanner.mll (hunk 6 above)
Comment (jibben) [reply src/scanner.mll:19]: We removed single line comments from spec

File: src/scanner.mll (hunk 3 above)
Comment (jibben) [reply src/scanner.mll:51]: we don't have a `func` keyword?

Comment (Howon) [reply src/scanner.mll:51]: lets use def now

Comment (jibben) [reply src/scanner.mll:51]: we are missing `let` though

File: src/scanner.mll [hunk 7]
```diff
@@ -23,30 +34,68 @@ rule token = parse
...
+    | '&'                   { BITWISE_AND }
+    | '|'                   { BITWISE_OR }
+    | '^'                   { BITWISE_XOR }
+    | '~'                   { BITWISE_NOT }
+    | ">>"                  { BITSHIFT_RIGHT }
+    | "<<"                  { BITSHIFT_LEFT }
+
+    (* functions *)
+    | ':'                   { FUNCTION_TYPE }
+    | "->"                  { ARROW }
```
Comment (Howon) [inline src/scanner.mll:50]: this is for key -> value association in MAPs

File: src/scanner.mll [hunk 8]
```diff
@@ -23,30 +34,68 @@ rule token = parse
...
+    | "def"                 { DEF }
+
+    (* literals *)
     | ['0'-'9']+ as lxm { LITERAL(int_of_string lit) }
+    | ['0'-'9']+'.'['0'-'9']* as lxm { DOUBLE_LIT(float_of_string lxm)}
+    | '\"' ([^'\"']* as lxm) '\"' { STRING_LIT(lxm) }
+    | '\'' (('a'-'z'|'A'-'Z')|'\\'['\\' '*' 'n' 'r' 't' '"' '''] as lxm) '\'' { CHAR_LIT(lxm) }
+
+    (* identifiers *)
     | ['a' - 'z' 'A' - 'Z']['a' - 'z' 'A' - 'Z' '0' - '9' '_']* as lxm { ID(lxm) }
```
Comment (Howon) [inline src/scanner.mll]: not sure if we need '_'

Comment (Howon) [inline src/scanner.mll]: lets make ie camel case only

**Comment by Howon** [comment]
change your branch names plz

**Review by aethos — APPROVED** [review]
let's merge this

**Comment by Howon** [comment]
Please add 'receive' keyword

File: src/scanner.mll [hunk 9]
```diff
@@ -3,18 +3,29 @@
 rule token = parse
     | [' ' '\t' '\r' '\n']  { token lexbuf }
     | "/*"                  { comment lexbuf }
+
+    (* braces/parens/brackets *)
```
Comment (anthonyholley) [inline src/scanner.mll:7]: I'm guessing we need to deal with the optional parens too: <>

**Review by anthonyholley — COMMENTED** [review]
need <> parens for options?

**Pull request merged**